flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreITCase.java

```diff
@@ -283,6 +286,54 @@ public void testConcurrentCheckpointOperations() throws Exception {
 		recoveredTestCheckpoint.awaitDiscard();
 	}
 
+	/**
+	 * FLINK-17073 tests that there is no request triggered when there are too many checkpoints
+	 * waiting to clean and that it resumes when the number of waiting checkpoints as gone below
+	 * the threshold.
+	 *
+	 */
+	@Test
+	public void testChekpointingPausesAndResumeWhenTooManyCheckpoints() throws Exception{
+		ManualClock clock = new ManualClock();
+		clock.advanceTime(1, TimeUnit.DAYS);
+		int maxCleaningCheckpoints = 1;
+		CheckpointsCleaner checkpointsCleaner = new CheckpointsCleaner();
+		CheckpointRequestDecider checkpointRequestDecider =  new CheckpointRequestDecider(maxCleaningCheckpoints, unused ->{}, clock, 1, new AtomicInteger(0)::get, checkpointsCleaner::getNumberOfCheckpointsToClean);
+
+		final int maxCheckpointsToRetain = 1;
+		Executors.PausableThreadPoolExecutor executor = Executors.pausableExecutor();
```

Comment:
I think `ManuallyTriggeredScheduledExecutor` would suffice here.
Did you consider using it?

Reply:
I was unaware of this class. Thanks for pointing out ! Yes the public API of this class fits the ITest needs.